File: C++/Level3/베스트앨범.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

struct info {
    int num;
    string genre;
    int play;
};

map<string, int> genre_plays;

bool comp(info i1, info i2) {

    if (genre_plays[i1.genre] > genre_plays[i2.genre]) return true;
    if (genre_plays[i1.genre] < genre_plays[i2.genre]) return false;

    if (i1.play > i2.play) return true;
    if (i1.play < i2.play) return false;

    return i1.num < i2.num;
}

vector<int> solution(vector<string> genres, vector<int> plays) {

    vector<int> answer;
    vector<info> songs;

    // 각각의 노래들의 번호와 정보들을 songs에 저장
    for (int i = 0; i < genres.size(); i++)
    {
        songs.push_back({ i, genres.at(i), plays.at(i) });
        genre_plays[genres.at(i)] += plays.at(i);
    }

    sort(songs.begin(), songs.end(), comp);

    for (int i = 0; i < songs.size(); i++)
    {
        if (i > 1 && songs.at(i).genre == songs.at(i - 1).genre && songs.at(i).genre == songs.at(i - 2).genre) continue;
        answer.push_back(songs.at(i).num);
    }

    return answer;
}

File: programmers/Java/Level3/베스트앨범.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Song {
	int num;
	int play;
	String genre;

	Song(int num, int play, String genre) {
		this.num = num;
		this.play = play;
		this.genre = genre;
	}
}

class Solution {
	public int[] solution(String[] genres, int[] plays) {
		int N = genres.length;
		Map<String, Integer> genrePlay = new HashMap<>(); // 장르별 총 재생 수 저장
		List<Song> songs = new ArrayList<>(); // 각 노래들 정보 저장
		List<Integer> album = new ArrayList<>(); // 앨범에 수록되는 노래 순서

		// 앨범 정보 등록
		for (int i = 0; i < N; i++) {
			genrePlay.put(genres[i], plays[i] + genrePlay.getOrDefault(genres[i], 0));
			songs.add(new Song(i, plays[i], genres[i]));
		}

		// 노래 수록 기준 정렬
		Collections.sort(songs,
				(o1, o2) -> genrePlay.get(o1.genre) != genrePlay.get(o2.genre)
						? genrePlay.get(o2.genre) - genrePlay.get(o1.genre)
						: (o1.play != o2.play) ? o2.play - o1.play : o1.num - o2.num);

		// 각 장르별로 2개의 노래만 수록
		for (int i = 0; i < N; i++) {
			if (i < 2 || !songs.get(i - 2).genre.equals(songs.get(i - 1).genre)
					|| !songs.get(i - 1).genre.equals(songs.get(i).genre)) {
				album.add(songs.get(i).num);
			}
		}

		// 결과 배열 만들고 반환
		int[] answer = new int[album.size()];
		for (int i = 0; i < album.size(); i++) {
			answer[i] = album.get(i);
		}
		return answer;
	}
}